src/approaches/grammar_search_invention_approach.py

```diff
@@ -820,14 +827,56 @@ def _evaluate_with_operators(self,
     @staticmethod
     def _get_refinement_prob(
             demo_atoms_sequence: Sequence[Set[GroundAtom]],
-            plan_atoms_sequence: Sequence[Collection[GroundAtom]]) -> float:
+            plan_atoms_sequence: Sequence[Set[GroundAtom]],
+            demo_multistep_effects: Set[FrozenSet[GroundAtom]]) -> float:
         """Estimate the probability that plan_atoms_sequence is refinable using
         the demonstration demo_atoms_sequence."""
-        # Make a soft assumption that the demonstrations are optimal.
+        # Make a soft assumption that the demonstrations are optimal,
+        # using a geometric distribution.
         demo_len = len(demo_atoms_sequence)
         plan_len = len(plan_atoms_sequence)
+        # The exponent is the difference in plan lengths.
+        exponent = abs(demo_len - plan_len)
+        if CFG.grammar_search_expected_nodes_include_suspicious_score:
+            # Handle suspicious effect scoring via unification.
+            num_suspicious_eff = 0
+            for i in range(len(plan_atoms_sequence) - 1):
+                for j in range(i + 1, len(plan_atoms_sequence)):
```

Comment:
interesting, this is a bit different from what I had in mind. I was thinking that we would specify a hyperparameter for what interval we want to use for the effects. for example, we'll just look at two-step effects.

this brings up a few questions:
1. how much slower is this way? one concern is that the effects could be quite large as you go further into the sequence, so unification could get slow.
2. are there examples that we know about where >2 step effects will help?
3. if we stick with this way, should we still compare 2-step demo effects with 2-step plan effects, then 3-step demo effects with 3-step plan effects, and so on? maybe not? i'm not sure.
4. should we skip 1-step effects because they should never be suspicious (because of our operator learning) except for min_data_for_nsrts>0?

Reply:
There's a lot of variants we can try. We just don't have the ability to test anything meaningfully so I just wanted to put this code in, but when we actually have a domain where these suspicious transitions matter, we can definitely try these variants.
1. See below comment for general speed, but yeah it would be a lot faster if you just do the hyperparameter because your loops would be O(n) not O(n^2) which matters a lot for painting where the demos are long
2. We don't even have examples where any of this helps right now, so TBD
3. Same, no idea, it comes down to experimentation
4. No, we shouldn't skip them because of min_data_for_nsrts, but I did check that nothing gets counted as suspicious if you limit to 1-step effects, likely bc min_data_for_nsrts isn't filtering anything out right now